src/redisearch_rs/trie_rs/src/trie.rs

```diff
@@ -0,0 +1,844 @@
+            .binary_search_by_key(&first_byte, |child| child.first_byte)
+            .ok()
+    }
+
+    /// Remove a child at the given index.
+    fn remove(&mut self, index: usize) -> Node<Data> {
+        self.0.remove(index).node
+    }
+
+    /// Check whether the children vector has only a single child.
+    fn is_single(&self) -> bool {
+        self.0.len() == 1
+    }
+
+    /// Check whether the children vector is empty.
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+/// The reference to the child node held inside its parent node.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct ChildRef<Data> {
+    /// The first byte of the value that this child node holds.
+    first_byte: c_char,
+    /// The reference to the actual child node.
+    node: Node<Data>,
+}
+
+/// Iterates over the entries of a [`TrieMap`] in lexicographical order
+/// of the keys.
+pub struct Iter<'tm, Data> {
+    /// Stack of nodes and whether they have been visited.
+    stack: Vec<(&'tm Node<Data>, bool)>,
+    /// Labels of the parent nodes, used to reconstruct the key.
+    prefixes: Vec<&'tm [c_char]>,
+}
+
+impl<'tm, Data> Iter<'tm, Data> {
+    /// Creates a new iterator over the entries of a [`TrieMap`].
+    fn new(root: Option<&'tm Node<Data>>) -> Self {
+        Self {
+            stack: root.into_iter().map(|node| (node, false)).collect(),
+            prefixes: Vec::new(),
+        }
+    }
+}
+
+impl<'tm, Data: fmt::Debug> Iterator for Iter<'tm, Data> {
+    type Item = (Vec<c_char>, &'tm Data);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (node, was_visited) = self.stack.pop()?;
+
+        if !was_visited {
+            let data = node.data.as_ref();
+            self.stack.push((node, true));
+
+            for child in node.children.0.iter().rev() {
+                self.stack.push((&child.node, false));
+            }
+
+            self.prefixes.push(&node.label);
+            if let Some(data) = data {
+                // Combine the labels of the parent nodes and the current node,
+                // thereby reconstructing the key.
+                let label = self
+                    .prefixes
+                    .iter()
+                    .flat_map(|p| p.iter())
+                    .copied()
+                    .collect();
+                return Some((label, data));
+            }
+        } else {
+            self.prefixes.pop();
+        }
+        self.next()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use proptest::prelude::proptest;
+
+    use super::*;
+
+    trait ToCCharArray<const N: usize> {
+        /// Convenience method to convert a byte array to a C-compatible character array.
+        fn c_chars(self) -> [c_char; N];
+    }
+
+    impl<const N: usize> ToCCharArray<N> for [u8; N] {
+        fn c_chars(self) -> [c_char; N] {
+            self.map(|b| b as c_char)
+        }
+    }
+
+    /// Forwards to `insta::assert_debug_snapshot!`,
+    /// but is disabled in Miri.
```

Comment:
Let's capture why we need to disable it for `miri`.

Reply:
Fixed in 4ca45094e575cd88387bc59e3fa7918201237a8e.